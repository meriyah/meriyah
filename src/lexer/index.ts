export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
