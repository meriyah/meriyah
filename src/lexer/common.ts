import { Chars } from '../chars';
import { Flags } from '../common';
import { type Parser } from '../parser/parser';
import { Token } from '../token';

export const enum LexerState {
  None = 0,
  NewLine = 1 << 0,
  LastIsCR = 1 << 2,
}

export const enum NumberKind {
  ImplicitOctal = 1 << 0,
  Binary = 1 << 1,
  Octal = 1 << 2,
  Hex = 1 << 3,
  Decimal = 1 << 4,
  NonOctalDecimal = 1 << 5,
  Float = 1 << 6,
  ValidBigIntKind = 1 << 7,
  DecimalNumberKind = Decimal | NonOctalDecimal,
}

/**
 * Advances this lexer's current index.
 *
 * @param parser The parser instance
 */
export function advanceChar(parser: Parser): number {
  parser.column++;
  return (parser.currentChar = parser.source.charCodeAt(++parser.index));
}

/**
 * Consumes possible surrogate pair, return the merged code point or 0
 *
 * @param parser The parser instance
 */
export function consumePossibleSurrogatePair(parser: Parser): number {
  // See: https://tc39.github.io/ecma262/#sec-ecmascript-language-types-string-type
  const hi = parser.currentChar;
  // Not surrogate
  if ((hi & 0xfc00) !== Chars.LeadSurrogateMin) return Chars.Null;
  const lo = parser.source.charCodeAt(parser.index + 1);
  // Unpaired surrogate
  if ((lo & 0xfc00) !== Chars.TrailSurrogateMin) return Chars.Null;
  // Merged code point
  return Chars.NonBMPMin + ((hi & 0x3ff) << 10) + (lo & 0x3ff);
}

/**
 * Use to consume a line feed instead of `scanNewLine`.
 */
export function consumeLineFeed(parser: Parser, state: LexerState): void {
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.flags |= Flags.NewLine;
  if ((state & LexerState.LastIsCR) === 0) {
    parser.column = 0;
    parser.line++;
  }
}

export function scanNewLine(parser: Parser): void {
  parser.flags |= Flags.NewLine;
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.column = 0;
  parser.line++;
}

// ECMA-262 11.2 White Space
// gC=Zs, U+0009, U+000B, U+000C, U+FEFF
export function isExoticECMAScriptWhitespace(ch: number): boolean {
  return (
    ch === Chars.NonBreakingSpace ||
    ch === Chars.ZeroWidthNoBreakSpace ||
    ch === Chars.NextLine ||
    ch === Chars.Ogham ||
    (ch >= Chars.EnQuad && ch <= Chars.ZeroWidthSpace) ||
    ch === Chars.NarrowNoBreakSpace ||
    ch === Chars.MathematicalSpace ||
    ch === Chars.IdeographicSpace ||
    ch === Chars.ThinSpace ||
    ch === Chars.ByteOrderMark
  );
}

/**
 * Converts a value to a hex value
 *
 * @param code CodePoint
 */
export function toHex(code: number): number {
  return code < Chars.UpperA ? code - Chars.Zero : (code - Chars.UpperA + 10) & 0xf;
}

/**
 * Converts a token to a string representation
 *
 * @param t Token
 */
export function convertTokenType(t: Token): string {
  switch (t) {
    case Token.NumericLiteral:
      return 'NumericLiteral';
    case Token.StringLiteral:
      return 'StringLiteral';
    case Token.FalseKeyword:
    case Token.TrueKeyword:
      return 'BooleanLiteral';
    case Token.NullKeyword:
      return 'NullLiteral';
    case Token.RegularExpression:
      return 'RegularExpression';
    case Token.TemplateContinuation:
    case Token.TemplateSpan:
    case Token.Template:
      return 'TemplateLiteral';
    default:
      if ((t & Token.IsIdentifier) === Token.IsIdentifier) return 'Identifier';
      if ((t & Token.Keyword) === Token.Keyword) return 'Keyword';

      return 'Punctuator';
  }
}
