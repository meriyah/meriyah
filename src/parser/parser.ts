import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
