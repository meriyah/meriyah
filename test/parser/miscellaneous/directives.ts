import { Context } from '../../../src/common';
import { pass, fail } from '../../test-utils';
import * as t from 'assert';
import { parseSource } from '../../../src/parser';

describe('Miscellaneous - Directives', () => {
  for (const arg of [
    '"\\1;" "use strict";',
    '"\\2;" "use strict";',
    '"\\3;" "use strict";',
    '"\\4;" "use strict";',
    '"\\5;" "use strict";',
    '"\\6;" "use strict";',
    '"\\7;" "use strict";',
    '"\\8;" "use strict";',
    '"\\9;" "use strict";',
    '"use strict"; function f(){"\\1";}',
    '"\\1;" "use strict"; null',
    '"use strict"; with (a) b = c;',
    '"use strict"; "\\1;"',
    '"use strict"; "\\2;"',
    '"use strict"; "\\3;"',
    '"use strict"; "\\4;"',
    '"use strict"; "\\5;"',
    '"use strict"; "\\6;"',
    '"use strict"; "\\7;"',
    '"use strict"; "\\8;"',
    '"use strict"; "\\9;"',
    '"use strict"; "\\1;" null',
    '"random\\x0\nnewline"',
    '"random\\u\nnewline"',
    '"random\\u0\nnewline"',
    '"random\\ua\u2029newline"',
    '"random\\ua\rnewline"',
    '"random\\u0a\nnewline"',
    '"random\\u000\nnewline"',
    '"random\\u00a\nnewline"',
    '"random\\u{0\nnewline"',
    '"random\\u{a\nnewline"',
    "'random\\x foo'",
    '"random\\u{a\rnewline"',
    "'random\\u foo'",
    "'random\\u0 foo'",
    "'random\\u00 foo'",
    "'random\\u0a foo'",
    "'random\\x0\\ foo'",
    "'random\\ua\\ foo'",
    "'random\\x0\\ foo'",
    "'random\\u0a\\ foo'",
    "'random\\xx foo'",
    "'random\\u00a\\ foo'",
    "'random\\uax foo'",
    "'random\\u0au foo'",
    '"use strict" ++',
    '"use strict" \\1',
    '"use strict" "\\1"',
    '"use strict"; "\\1";',
    '"use strict" ++',
    `function foo() { "use strict"; with (a) b = c; }`,
    '"use strict"; function foo() { with (a) b = c; }',
    '"use strict"; function hello() { "\\000"; }',
    '"use strict"; function hello() { "\\00"; }',
    '"use strict"; function hello() { "\\0123"; }',
    'function hello() { "use strict"; "\\000"; }',
    'function hello() { "use strict"; "\\00"; }',
    'function hello() { "use strict"; "\\0123"; }',
    'function hello("\\000008") { "use strict"; }',
    ` function fun() {
                "use strict";
                       var public = 1;
            }`,
    ` function fun() {
              "use strict"
                     var public = 1;
          }`
  ]) {
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, undefined, Context.None);
      });
    });

    it(`/* comment in front */ ${arg}`, () => {
      t.throws(() => {
        parseSource(`/* comment in front */ ${arg}`, undefined, Context.None);
      });
    });

    it(`function foo() { ${arg} }`, () => {
      t.throws(() => {
        parseSource(`function foo() { ${arg} }`, undefined, Context.Strict | Context.Module);
      });
    });
  }

  for (const arg of [
    '("use strict")',
    '"\\n\\r\\t\\v\\b\\f\\\\\\\'\\"\\0"',
    '"use some future directive"',
    '"use some future directive";',
    '"use some future directive"; "use strict";',
    '"Hello\\312World"',
    '"use strict"',
    "'use\\x20strict'",
    '"use\\x20strict"',
    "'use asm'",
    "'use asm'; 'use strict'",
    "'use asm' \n 'use strict'",
    '"use asm" \n "use strict"',
    "'use asm' \r 'use strict'",
    '"use asm" \r "use strict"',
    "'use asm' \r\n 'use strict'",
    '"use asm" \r\n "use strict"',
    "'use asm' \u2028 'use strict'",
    '"use asm" \u2028 "use strict"',
    "'use asm' \u2029 'use strict'",
    '"use asm" \u2029 "use strict"',
    'function foo() { "use \\u0020strict"; with (a) b = c; }',
    '"use \\u0020strict"; function foo() { with (a) b = c; }',
    '"use strict"\n foo',
    "'use strict'; foo",
    'function foo() { "use strict"\n bar }',
    '!function foo() { "use strict"\n bar }',
    '() => { "use strict"\n foo }',
    '() => "use strict"',
    '({ wrap() { "use strict"; foo } })',
    '"\\u0075se strict"',
    '"use asm"; "use strict"; foo',
    'function wrap() { "use asm"; "use strict"; foo }',
    '"use strict"; foo; "use asm"',
    'function wrap() { "use asm"; foo; "use strict" }',
    '{ "use strict"; }',
    'function wrap() { { "use strict" } foo }',
    '("use strict"); foo',
    'function wrap() { ("use strict"); foo }',
    'function a() { "use strict" } "use strict"; foo',
    'try { "use strict"; var public = 1; } catch (e) {}',
    '"use asm" \u2029 "use strict"',
    '"\\n\\r\\t\\v\\b\\f\\\\\\\'\\"\\0"',
    '"Hello\\312World"',
    `"use strict"; + 1`,
    `function wrap() { "use strict"\n foo }`,
    `"\\u0075se strict"`,
    'function wrap() { { "use strict" } foo }',
    '"Hello\\0World"',
    '("use strict"); foo = 42;',
    '("use strict"); eval = 42;',
    '"USE STRICT";  var public = 1;',
    'function wrap() { "use asm"; "use strict"; foo }',
    '{ "use strict"; }',
    'function a() { "use strict" } "use strict"; foo;',
    'function foo() { "use \\u0020strict"; with (a) b = c; }',
    '"use \\u0020strict"; with (a) b = c;',
    `function foo()
    {
       "bogus directive";
       "use strict";
       return (this === undefined);
    }`,
    '"use strict", "Hello\\312World"',
    '"use strict" + "Hello\\312World"',
    '"use strict", "Hello\\312World"',
    '"use strict", "Hello\\312World"',
    'function foo() { "use strict" .foo }',
    '"use strict"; "\\0";',
    '"\\0"; "use strict";',
    'function a() {"use strict"; "\\0";}',
    'function a() {"\\0"; "use strict";}'
  ]) {
    it(`/* comment in front */ ${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`/* comment in front */ ${arg}`, undefined, Context.OptionsRaw);
      });
    });

    it(`/* comment in front */ ${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`/* comment in front */ ${arg}`, undefined, Context.OptionsWebCompat | Context.OptionsRaw);
      });
    });
  }

  for (const arg of ['.foo', '[foo]', '()', '`x`', ' + x', '/f', '/f/g']) {
    t.throws(() => {
      parseSource(`function f(){ "use strict" \n /* suffix = */   ${arg} ; eval = 1; }`, undefined, Context.Strict);
    });
  }

  for (const arg of ['foo', '++x', '--x', 'function f(){}', '{x}', ';', '25', 'true']) {
    t.throws(() => {
      parseSource(`function f(){ "use strict" \n  ${arg} ; eval = 1; }`, undefined, Context.None);
    });
  }

  fail('Miscellaneous - Directives (fail)', [
    ['"use strict"; var static;', Context.None],
    ['\\u0061sync function f(){}', Context.None],
    ['"use strict" "Hello\\312World"', Context.None],
    ['"use strict" \n "Hello\\312World"', Context.None],
    ['function a() { "use strict" "Hello\\312World" }', Context.None],
    ['function a() { "use strict" \n "Hello\\312World" }', Context.None]
  ]);

  pass('Miscellaneous - Directives (pass)', [
    [
      '"use strict" + "Hello\\312World"',
      Context.Module | Context.OptionsRanges | Context.OptionsRaw,
      {
        body: [
          {
            expression: {
              left: {
                raw: '"use strict"',
                start: 0,
                end: 12,
                range: [0, 12],
                type: 'Literal',
                value: 'use strict'
              },
              operator: '+',
              right: {
                raw: '"Hello\\312World"',
                start: 15,
                end: 31,
                range: [15, 31],
                type: 'Literal',
                value: 'HelloÊWorld'
              },
              start: 0,
              end: 31,
              range: [0, 31],
              type: 'BinaryExpression'
            },
            start: 0,
            end: 31,
            range: [0, 31],
            type: 'ExpressionStatement'
          }
        ],
        sourceType: 'module',
        start: 0,
        end: 31,
        range: [0, 31],
        type: 'Program'
      }
    ],
    [
      '("use strict"); foo = 42;',
      Context.Module | Context.OptionsRanges | Context.OptionsRaw,
      {
        body: [
          {
            expression: {
              raw: '"use strict"',
              start: 1,
              end: 13,
              range: [1, 13],
              type: 'Literal',
              value: 'use strict'
            },
            start: 0,
            end: 15,
            range: [0, 15],
            type: 'ExpressionStatement'
          },
          {
            expression: {
              left: {
                name: 'foo',
                start: 16,
                end: 19,
                range: [16, 19],
                type: 'Identifier'
              },
              operator: '=',
              right: {
                raw: '42',
                start: 22,
                end: 24,
                range: [22, 24],
                type: 'Literal',
                value: 42
              },
              start: 16,
              end: 24,
              range: [16, 24],
              type: 'AssignmentExpression'
            },
            start: 16,
            end: 25,
            range: [16, 25],
            type: 'ExpressionStatement'
          }
        ],
        sourceType: 'module',
        start: 0,
        end: 25,
        range: [0, 25],
        type: 'Program'
      }
    ],
    [
      '"use strict", "Hello\\312World"',
      Context.None | Context.OptionsRanges | Context.OptionsRaw,
      {
        body: [
          {
            expression: {
              expressions: [
                {
                  start: 0,
                  end: 12,
                  range: [0, 12],
                  type: 'Literal',
                  value: 'use strict',
                  raw: '"use strict"'
                },
                {
                  start: 14,
                  end: 30,
                  range: [14, 30],
                  type: 'Literal',
                  value: 'HelloÊWorld',
                  raw: '"Hello\\312World"'
                }
              ],
              start: 0,
              end: 30,
              range: [0, 30],
              type: 'SequenceExpression'
            },
            start: 0,
            end: 30,
            range: [0, 30],
            type: 'ExpressionStatement'
          }
        ],
        sourceType: 'script',
        start: 0,
        end: 30,
        range: [0, 30],
        type: 'Program'
      }
    ],
    [
      '"use asm" \n "use strict"',
      Context.None | Context.OptionsRanges | Context.OptionsRaw,
      {
        body: [
          {
            expression: {
              start: 0,
              end: 9,
              range: [0, 9],
              type: 'Literal',
              value: 'use asm',
              raw: '"use asm"'
            },
            directive: 'use asm',
            start: 0,
            end: 9,
            range: [0, 9],
            type: 'ExpressionStatement'
          },
          {
            expression: {
              start: 12,
              end: 24,
              range: [12, 24],
              type: 'Literal',
              value: 'use strict',
              raw: '"use strict"'
            },
            directive: 'use strict',
            start: 12,
            end: 24,
            range: [12, 24],
            type: 'ExpressionStatement'
          }
        ],
        sourceType: 'script',
        start: 0,
        end: 24,
        range: [0, 24],
        type: 'Program'
      }
    ],
    [
      '"use strict"; + 1',
      Context.None | Context.OptionsRanges,
      {
        type: 'Program',
        sourceType: 'script',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'Literal',
              value: 'use strict',
              start: 0,
              end: 12,
              range: [0, 12]
            },
            directive: 'use strict',
            start: 0,
            end: 13,
            range: [0, 13]
          },
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'UnaryExpression',
              operator: '+',
              argument: {
                type: 'Literal',
                value: 1,
                start: 16,
                end: 17,
                range: [16, 17]
              },
              prefix: true,
              start: 14,
              end: 17,
              range: [14, 17]
            },
            start: 14,
            end: 17,
            range: [14, 17]
          }
        ],
        start: 0,
        end: 17,
        range: [0, 17]
      }
    ],
    [
      '("use strict"); foo = 42;',
      Context.None | Context.OptionsRanges,
      {
        type: 'Program',
        sourceType: 'script',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'Literal',
              value: 'use strict',
              start: 1,
              end: 13,
              range: [1, 13]
            },
            start: 0,
            end: 15,
            range: [0, 15]
          },
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'AssignmentExpression',
              left: {
                type: 'Identifier',
                name: 'foo',
                start: 16,
                end: 19,
                range: [16, 19]
              },
              operator: '=',
              right: {
                type: 'Literal',
                value: 42,
                start: 22,
                end: 24,
                range: [22, 24]
              },
              start: 16,
              end: 24,
              range: [16, 24]
            },
            start: 16,
            end: 25,
            range: [16, 25]
          }
        ],
        start: 0,
        end: 25,
        range: [0, 25]
      }
    ],
    [
      '"\\u0061b"\n"c\\u0064"',
      Context.None,
      {
        type: 'Program',
        sourceType: 'script',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'Literal',
              value: 'ab'
            },
            directive: '\\u0061b'
          },
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'Literal',
              value: 'cd'
            },
            directive: 'c\\u0064'
          }
        ]
      }
    ],
    [
      '"\\u0061b"\n"c\\u0064"',
      Context.Module,
      {
        type: 'Program',
        sourceType: 'module',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'Literal',
              value: 'ab'
            },
            directive: '\\u0061b'
          },
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'Literal',
              value: 'cd'
            },
            directive: 'c\\u0064'
          }
        ]
      }
    ]
  ]);
});
