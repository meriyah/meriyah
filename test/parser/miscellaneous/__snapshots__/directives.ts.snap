// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Miscellaneous - Directives > Miscellaneous - Directives (fail) > "use strict" 
 "Hello\\312World" 1`] = `
"SyntaxError [2:1-2:17]: Octal literals are not allowed in strict mode
  1 | "use strict" 
> 2 |  "Hello\\312World"
    |  ^^^^^^^^^^^^^^^^ Octal literals are not allowed in strict mode"
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (fail) > "use strict" "Hello\\312World" 1`] = `
"SyntaxError [1:13-1:29]: Unexpected token: 'string'
> 1 | "use strict" "Hello\\312World"
    |              ^^^^^^^^^^^^^^^^ Unexpected token: 'string'"
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (fail) > "use strict" + "Hello\\312World" 1`] = `
"SyntaxError [1:15-1:22]: Octal escape sequences are not allowed in strict mode
> 1 | "use strict" + "Hello\\312World"
    |                ^^^^^^^ Octal escape sequences are not allowed in strict mode"
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (fail) > "use strict"; var static; 1`] = `
"SyntaxError [1:18-1:24]: Unexpected strict mode reserved word
> 1 | "use strict"; var static;
    |                   ^^^^^^ Unexpected strict mode reserved word"
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (fail) > \\u0061sync function f(){} 1`] = `
"SyntaxError [1:11-1:19]: Unexpected token: 'function'
> 1 | \\u0061sync function f(){}
    |            ^^^^^^^^ Unexpected token: 'function'"
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (fail) > function a() { "use strict" 
 "Hello\\312World" } 1`] = `
"SyntaxError [1:15-2:17]: Octal literals are not allowed in strict mode
> 1 | function a() { "use strict" 
    |                ^^^^^^^^^^^^^
> 2 |  "Hello\\312World" }
    | ^^^^^^^^^^^^^^^^^^ Octal literals are not allowed in strict mode"
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (fail) > function a() { "use strict" "Hello\\312World" } 1`] = `
"SyntaxError [1:28-1:44]: Unexpected token: 'string'
> 1 | function a() { "use strict" "Hello\\312World" }
    |                             ^^^^^^^^^^^^^^^^ Unexpected token: 'string'"
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (pass) > "\\u0061b"
"c\\u0064" 1`] = `
{
  "body": [
    {
      "directive": "\\u0061b",
      "expression": {
        "type": "Literal",
        "value": "ab",
      },
      "type": "ExpressionStatement",
    },
    {
      "directive": "c\\u0064",
      "expression": {
        "type": "Literal",
        "value": "cd",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (pass) > "\\u0061b"
"c\\u0064" 2`] = `
{
  "body": [
    {
      "directive": "\\u0061b",
      "expression": {
        "type": "Literal",
        "value": "ab",
      },
      "type": "ExpressionStatement",
    },
    {
      "directive": "c\\u0064",
      "expression": {
        "type": "Literal",
        "value": "cd",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (pass) > "use asm" 
 "use strict" 1`] = `
{
  "body": [
    {
      "directive": "use asm",
      "end": 9,
      "expression": {
        "end": 9,
        "range": [
          0,
          9,
        ],
        "raw": ""use asm"",
        "start": 0,
        "type": "Literal",
        "value": "use asm",
      },
      "range": [
        0,
        9,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
    {
      "directive": "use strict",
      "end": 24,
      "expression": {
        "end": 24,
        "range": [
          12,
          24,
        ],
        "raw": ""use strict"",
        "start": 12,
        "type": "Literal",
        "value": "use strict",
      },
      "range": [
        12,
        24,
      ],
      "start": 12,
      "type": "ExpressionStatement",
    },
  ],
  "end": 24,
  "range": [
    0,
    24,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (pass) > "use strict" + "Hello\\312World" 1`] = `
{
  "body": [
    {
      "end": 31,
      "expression": {
        "end": 31,
        "left": {
          "end": 12,
          "range": [
            0,
            12,
          ],
          "raw": ""use strict"",
          "start": 0,
          "type": "Literal",
          "value": "use strict",
        },
        "operator": "+",
        "range": [
          0,
          31,
        ],
        "right": {
          "end": 31,
          "range": [
            15,
            31,
          ],
          "raw": ""Hello\\312World"",
          "start": 15,
          "type": "Literal",
          "value": "HelloÊWorld",
        },
        "start": 0,
        "type": "BinaryExpression",
      },
      "range": [
        0,
        31,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 31,
  "range": [
    0,
    31,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (pass) > "use strict", "Hello\\312World" 1`] = `
{
  "body": [
    {
      "end": 30,
      "expression": {
        "end": 30,
        "expressions": [
          {
            "end": 12,
            "range": [
              0,
              12,
            ],
            "raw": ""use strict"",
            "start": 0,
            "type": "Literal",
            "value": "use strict",
          },
          {
            "end": 30,
            "range": [
              14,
              30,
            ],
            "raw": ""Hello\\312World"",
            "start": 14,
            "type": "Literal",
            "value": "HelloÊWorld",
          },
        ],
        "range": [
          0,
          30,
        ],
        "start": 0,
        "type": "SequenceExpression",
      },
      "range": [
        0,
        30,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 30,
  "range": [
    0,
    30,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (pass) > "use strict"; + 1 1`] = `
{
  "body": [
    {
      "directive": "use strict",
      "end": 13,
      "expression": {
        "end": 12,
        "range": [
          0,
          12,
        ],
        "start": 0,
        "type": "Literal",
        "value": "use strict",
      },
      "range": [
        0,
        13,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
    {
      "end": 17,
      "expression": {
        "argument": {
          "end": 17,
          "range": [
            16,
            17,
          ],
          "start": 16,
          "type": "Literal",
          "value": 1,
        },
        "end": 17,
        "operator": "+",
        "prefix": true,
        "range": [
          14,
          17,
        ],
        "start": 14,
        "type": "UnaryExpression",
      },
      "range": [
        14,
        17,
      ],
      "start": 14,
      "type": "ExpressionStatement",
    },
  ],
  "end": 17,
  "range": [
    0,
    17,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (pass) > ("use strict"); foo = 42; 1`] = `
{
  "body": [
    {
      "end": 15,
      "expression": {
        "end": 13,
        "range": [
          1,
          13,
        ],
        "raw": ""use strict"",
        "start": 1,
        "type": "Literal",
        "value": "use strict",
      },
      "range": [
        0,
        15,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
    {
      "end": 25,
      "expression": {
        "end": 24,
        "left": {
          "end": 19,
          "name": "foo",
          "range": [
            16,
            19,
          ],
          "start": 16,
          "type": "Identifier",
        },
        "operator": "=",
        "range": [
          16,
          24,
        ],
        "right": {
          "end": 24,
          "range": [
            22,
            24,
          ],
          "raw": "42",
          "start": 22,
          "type": "Literal",
          "value": 42,
        },
        "start": 16,
        "type": "AssignmentExpression",
      },
      "range": [
        16,
        25,
      ],
      "start": 16,
      "type": "ExpressionStatement",
    },
  ],
  "end": 25,
  "range": [
    0,
    25,
  ],
  "sourceType": "module",
  "start": 0,
  "type": "Program",
}
`;

exports[`Miscellaneous - Directives > Miscellaneous - Directives (pass) > ("use strict"); foo = 42; 2`] = `
{
  "body": [
    {
      "end": 15,
      "expression": {
        "end": 13,
        "range": [
          1,
          13,
        ],
        "start": 1,
        "type": "Literal",
        "value": "use strict",
      },
      "range": [
        0,
        15,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
    {
      "end": 25,
      "expression": {
        "end": 24,
        "left": {
          "end": 19,
          "name": "foo",
          "range": [
            16,
            19,
          ],
          "start": 16,
          "type": "Identifier",
        },
        "operator": "=",
        "range": [
          16,
          24,
        ],
        "right": {
          "end": 24,
          "range": [
            22,
            24,
          ],
          "start": 22,
          "type": "Literal",
          "value": 42,
        },
        "start": 16,
        "type": "AssignmentExpression",
      },
      "range": [
        16,
        25,
      ],
      "start": 16,
      "type": "ExpressionStatement",
    },
  ],
  "end": 25,
  "range": [
    0,
    25,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;
